lib/ngtcp2_conn.c

```diff
@@ -6089,7 +6089,7 @@ static int conn_shutdown_stream_read(ngtcp2_conn *conn, ngtcp2_strm *strm,
     return 0;
   }
 
-  strm->flags |= NGTCP2_STRM_FLAG_STOP_SENDING;
+  strm->flags |= NGTCP2_STRM_FLAG_STOP_SENDING | NGTCP2_STRM_FLAG_RECV_RST;
```

Comment:
This change ensure the stream can be closed correctly. If NGTCP2_STRM_FLAG_RECV_RST is not set, the stream can not be closed by ngtcp2_conn_close_stream_if_shut_rdwr(). NGTCP2_STRM_FLAG_RECV_RST will be set by conn_recv_rst_stream(), but peer may not send RST_STREAM frame when receiving STOP_SENDING frame because the stream may has been closed by peer. If peer did not send RST_STREAM frame, the NGTCP2_STRM_FLAG_RECV_RST will not be set, and the related stream can not be closed. "ngtcp2_rob_first_gap_offset(&strm->rob) == strm->last_rx_offset" can be always false because of one loss packet and NGTCP2_STRM_FLAG_STOP_SENDING flag.

Reply:
In QUIC, peer has to close its send stream either STREAM with fin bit set or RST_STREAM.
Even if peer receives STOP_SENDING, it has to do either sending STREAM with fin or RST_STREAM.  If peer does not do this, it is broken.

Reply:
If the related stream has been closed by peer, How can peer send RST_STREAM or fin ?

Reply:
What is "related" stream?
Bidirectional stream has states for both directions independently.  Only sending peer can close it.
If an endpoint receives RST_STREAM from peer, it has to send STREAM FIN or RST_STREAM to close its sending stream.  Receiving RST_STREAM does not mean both side of bidirectional stream are closed.